mmhuman3d/models/utils/SMPLX.py

```diff
@@ -0,0 +1,514 @@
+        self.num_expression_coeffs = num_expression_coeffs
+
+    def __call__(self, body_predictions, face_predictions):
+        body_predictions['pred_param']['jaw_pose'] = face_predictions['pred_param']['jaw_pose']
+        body_predictions['pred_param']['expression'] = face_predictions['pred_param']['expression'][:,:self.num_expression_coeffs]
+        return body_predictions
+
+
+def points_to_bbox(points, bbox_scale_factor: float = 1.0):
+
+    min_coords, _ = torch.min(points, dim=1)
+    xmin, ymin = min_coords[:, 0], min_coords[:, 1]
+    max_coords, _ = torch.max(points, dim=1)
+    xmax, ymax = max_coords[:, 0], max_coords[:, 1]
+
+    center = torch.stack(
+        [xmax + xmin, ymax + ymin], dim=-1) * 0.5
+
+    width = (xmax - xmin)
+    height = (ymax - ymin)
+
+    # Convert the bounding box to a square box
+    size = torch.max(width, height) * bbox_scale_factor
+
+    return center, size
+
+def get_crop_info(points, img_metas, scale_factor: float = 1.0, crop_size: int = 256):
+    device = points.device
+    dtype = points.dtype
+    batch_size = points.shape[0]
+    # Get the image to crop transformations and bounding box sizes
+    crop_transforms = []
+    img_bbox_sizes = []
+    for img_meta in img_metas:
+        crop_transforms.append(img_meta['crop_transform'])
+        img_bbox_sizes.append(img_meta['scale'].max())
+
+    img_bbox_sizes = torch.tensor(
+        img_bbox_sizes, dtype=dtype, device=device)
+
+    crop_transforms = torch.tensor(
+        crop_transforms, dtype=dtype, device=device)
+
+    crop_transforms = torch.cat([crop_transforms,
+        torch.tensor([0.0,0.0,1.0],dtype = dtype, device=device).expand([batch_size,1,3])
+        ], dim = 1)
+
+    inv_crop_transforms = torch.inverse(crop_transforms)
+
+    # center on the cropped body image
+    center_body_crop, bbox_size = points_to_bbox(
+        points, bbox_scale_factor=scale_factor)
+
+    orig_bbox_size = bbox_size / crop_size * img_bbox_sizes
+
+    # Compute the center of the crop in the original image
+    center = (torch.einsum(
+        'bij,bj->bi', [inv_crop_transforms[:, :2, :2], center_body_crop]) +
+        inv_crop_transforms[:, :2, 2])
+
+    return {'center': center.reshape(-1, 2),
+            'orig_bbox_size': orig_bbox_size,
+            # 'bbox_size': bbox_size.reshape(-1),
+            'inv_crop_transforms': inv_crop_transforms,
+            # 'center_body_crop': 2 * center_body_crop / (crop_size-1) - 1,
+            }
+
+def concat_images(images:List[torch.Tensor]):
+    sizes = [img.shape[1:] for img in images]
+    H, W = [max(s) for s in zip(*sizes)]
+    batch_size = len(images)
+    batched_shape = (batch_size, images[0].shape[0], H, W)
+    batched = torch.zeros(
+        batched_shape, device=images[0].device, dtype=images[0].dtype)
+    for ii, img in enumerate(images):
+        shape = img.shape
+        batched[ii, :shape[0], :shape[1], :shape[2]] = img
+    return batched
+
+def flip_pose(pose_rotmat):
```

Comment:
Can we use existing functions instead for `flip_pose`?

Reply:
No existing functions found.